include/seqan3/search/dream_index/binning_directory.hpp

```diff
@@ -0,0 +1,218 @@
+    binning_directory & operator=(binning_directory &&) = default;      //!< Defaulted.
+    ~binning_directory() = default;                                     //!< Defaulted.
+
+    /*!\brief Construct using number of bins, bitvector size and the number of hash functions.
+     * \param bins_     The number of bins.
+     * \param bits_     The bitvector size.
+     * \param num_hash_ The number of hash functions. Default 2.
+     */
+    binning_directory(size_t bins_, size_t bits_, size_t num_hash_ = 2)
+    {
+        assert(bins_ > 0);
+        assert(bits_ > 0);
+        assert(num_hash_ > 0);
+        if (bits_ < bins_)
+            throw std::logic_error{"There must be at least one bit per bin. Increase the current size of " +
+                                   std::to_string(bits_) + " to at least " + std::to_string(bins_) + "."};
+        bins = bins_;
+        num_hash = num_hash_;
+        data = sdsl::bit_vector(bits_);
+        bin_width   = (bins + 63) >> 6;
+        block_size  = bin_width << 6;
+        block_count = bits_ / block_size;
+
+        pre_hash.resize(num_hash);
+
+        for (size_t i = 0; i < num_hash; ++i)
+            pre_hash[i] = i ^ (15 * seed);
+    }
+    //!\}
+
+    /*!\brief Inserts a value into a specific bin.
+     * \param h   The raw hash value to process.
+     * \param bin The bin to insert into.
+     */
+    inline void set(size_t const h, size_t const bin)
+    {
+        assert(bin < bins);
+        // Sets respective positions in bitvector TODO SIMD?
+        std::ranges::for_each(pre_hash, [this, h, bin] (size_t const & val)
+        {
+            size_t idx = val * h;
+            hash_and_fit(idx);
+            idx += bin;
+            assert (idx < data.size());
+            data[idx] = 1;
+        });
+    }
+
+    /*!\brief Sets the number of bins.
+     * \param new_bins The new number of bins.
+     * \throws std::invalid_argument If passed number of bins is smaller than current number of bins.
+     * \attention The new number of new bins must be greater or equal to the current number of bins.
+     *
+     * The resulting binning_directory has an increased size proportional to the increase in the bin_width, e.g.
+     * resizing a binning_directory with 40 bins to 73 bins also increases the bin_width from 64 to 128 and hence the
+     * new binning_directory will be twice the size.
+     * This increase in size is necessary to avoid invalidating all computed hash functions.
+     * If you want to add more bins while keeping the size constant, you need to rebuild the binning_directory.
+     */
+    inline void resize(size_t const new_bins)
+    {
+        if (new_bins < bins)
+        {
+            throw std::invalid_argument{"The new number of new bins must be greater"
+                                        " or equal to the current number of bins."};
+        }
+
+        size_t bin_width_ = (new_bins + 63) >> 6;
```

Comment:
Can you add a comment what happens here?

EDIT: Also `bin_width_` is not very expressive. Maybe `new_bin_width`